File: LeetCode/Hard/0332-reconstruct-itinerary/0332-reconstruct-itinerary.java
class Solution {
    public List<String> findItinerary(List<List<String>> tickets) {
        List<String> result = new ArrayList<>();
        Map<String, PriorityQueue<String>> map = new HashMap<>();

        for (List<String> ticket : tickets) {
            String from = ticket.get(0);
            String to = ticket.get(1);
            
            map.computeIfAbsent(from, k -> new PriorityQueue<>()).add(to);
        }

        dfs(result, map, "JFK");
        return result;
    }

    private void dfs(List<String> result, Map<String, PriorityQueue<String>> map, String from) {
        while (map.containsKey(from) && !map.get(from).isEmpty()) {
            dfs(result, map, map.get(from).poll());
        }

        result.addFirst(from);
    }
}

File: LeetCode/Hard/0332-reconstruct-itinerary/0332-reconstruct-itinerary.kt
class Solution {
    fun findItinerary(tickets: List<List<String>>): List<String> {
        val result = mutableListOf<String>()
        val map = mutableMapOf<String, PriorityQueue<String>>()

        for (ticket in tickets) {
            map.getOrPut(ticket[0]) { PriorityQueue() }.add(ticket[1])
        }

        fun dfs(from: String) {
            while (map.containsKey(from) && map[from]!!.isNotEmpty()) {
                dfs(map[from]!!.poll())
            }

            result.add(from)
        }

        dfs("JFK")
        return result.reversed()
    }
}